### PR TITLE
AirfRANS: Fourier+5L/256d and 4L/320d — capacity scaling long run

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current winner uses 4L/256d (PR #2478, 0.2387). With the full 180-min budget, a larger model gets enough epochs to converge. Two directions to explore: (1) deeper — 5L/256d adds one more attention layer which may better capture boundary layer physics and improve the pressure channel which still dominates error; (2) wider — 4L/320d maintains depth but expands hidden dim. Both have the same architecture family as the winner but with more capacity. The 180-min budget should give 40+ epochs for either variant.

## Instructions

Run two parallel experiments on separate GPUs:

```bash
# Run 1: 5L/256d/4H (deeper)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 150 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --wandb_group roy-airfrans-capacity &

# Run 2: 4L/320d/4H (wider)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 150 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 320 \
  --model-heads 4 \
  --epochs 999 \
  --wandb_group roy-airfrans-capacity &

wait
```

**CRITICAL:** `--epochs 999` is required — the default in train.py is 2 epochs regardless of SENPAI_MAX_EPOCHS.

## Baseline

- **Primary metric:** `val_primary/surface_mse` (lower is better)
- **Current best (val):** 0.2387
- **Current best (test):** 0.2079
- **full_val/volume_mse:** 0.2933
- **Best PR:** #2478 (senku — Fourier + 4L/256d + no-EMA, 8 epochs, AdamW lr=5e-4, T_max=150)
- **Baseline W&B run:** vwb9teqa
- **Reproduce command:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 5e-4 --cosine_t_max 150 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999`

## Report

Please report for each run (5L/256d and 4L/320d):
- val_primary/surface_mse (final and per epoch)
- test_primary/surface_mse (final)
- Per-channel surface MSE breakdown: Ux, Uy, p, nut
- Total epochs completed
- Minutes per epoch (to gauge capacity vs throughput tradeoff)

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2500 because Superseded: roy reassigned to #2568 (AirfRANS lr=3e-4).

**Comment by tcapelle** [comment]
STUDENT roy:

## Results

### Run 1: 5L/256d/4H (deeper model) — W&B run: eiujqb9k

| Metric | Value | Baseline (4L/256d) | Change |
|--------|-------|-------------------|--------|
| **val_primary/surface_mse** | **0.0133** (epoch 73) | 0.2387 | **-94.4%** |
| **test_primary/surface_mse** | **0.0186** | 0.2079 | **-91.1%** |
| full_val/surface_mse_Ux | 2.01e-05 | — | — |
| full_val/surface_mse_Uy | 5.47e-06 | — | — |
| full_val/surface_mse_p | 0.0531 | — | — |
| full_val/surface_mse_nut | 2.02e-06 | — | — |
| full_val/volume_mse | 0.1027 | 0.2933 | -65.0% |
| full_test/surface_mse_p | 0.0743 | — | — |
| Epochs completed | 73 | 8 | — |
| Minutes per epoch | ~2.5 min | ~22 min | — |

Command:
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 \
  --cosine-t-max 150 --no-use-ema --enable-fourier --model-layers 5 \
  --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --agent roy --wandb-name "roy/airfrans-5L256d-deeper" \
  --wandb-group "roy-airfrans-capacity"
```

### Run 2: 4L/320d/4H (wider model) — W&B run: vyvrhvwv

| Metric | Value | Baseline (4L/256d) | Change |
|--------|-------|-------------------|--------|
| **val_primary/surface_mse** | **0.0150** (epoch 54) | 0.2387 | **-93.7%** |
| **test_primary/surface_mse** | **0.0387** | 0.2079 | **-81.4%** |
| full_val/surface_mse_Ux | 3.01e-05 | — | — |
| full_val/surface_mse_Uy | 6.20e-06 | — | — |
| full_val/surface_mse_p | 0.0600 | — | — |
| full_val/surface_mse_nut | 6.08e-06 | — | — |
| full_val/volume_mse | 0.1364 | 0.2933 | -53.5% |
| full_test/surface_mse_p | 0.1544 | — | — |
| Epochs completed | 66 | 8 | — |
| Minutes per epoch | ~2.7 min | ~22 min | — |

Command:
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 \
  --cosine-t-max 150 --no-use-ema --enable-fourier --model-layers 4 \
  --model-hidden-dim 320 --model-heads 4 --epochs 999 \
  --agent roy --wandb-name "roy/airfrans-4L320d-wider" \
  --wandb-group "roy-airfrans-capacity"
```

## What happened

**Both models massively outperformed the baseline** — the 180-min budget allowed 66-73 epochs vs the baseline's 8, and both models continued improving throughout.

**Run 1 (5L/256d deeper) is the winner** with val 0.0133 (-94.4%) and test 0.0186 (-91.1%). The deeper model had a slower start and larger LR-cycle spikes early on, but converged more steadily in later epochs. Its test metric (0.0186) is much better than Run 2's (0.0387), indicating better generalization.

**Run 2 (4L/320d wider)** achieved a very strong val of 0.0150 (-93.7%), but its test metric (0.0387) is significantly worse than its val, suggesting some overfitting. The wider model showed increasing training instability in later epochs, with large spikes during low-LR phases that weren't present earlier.

**Key observations:**
- Cosine LR with T_max=150 steps creates periodic cycling (~every 5 epochs). Best results oscillate between LR phases — low-LR epochs gave best results early, but later in training the pattern became less predictable.
- The pressure channel (surface_mse_p) drives the composite metric. It dropped from ~1.6 at epoch 1 to 0.053 (Run 1) and 0.060 (Run 2) at best epochs.
- Both models processed ~2.5 min/epoch — much faster than the baseline's 22 min/epoch. The 5L model actually ran faster than expected, completing 73 epochs vs the wider model's 66.
- Volume MSE also improved significantly: 0.1027 (Run 1) vs 0.2933 baseline (-65%).

## Suggested follow-ups

1. **Deeper model with more capacity**: Try 6L/256d or 5L/320d to see if adding even more depth or width continues the trend.
2. **Cosine schedule tuning**: T_max=150 causes rapid LR cycling that creates instability. A longer T_max (e.g., 500 or 1000) might give smoother convergence.
3. **Checkpoint best model**: Currently no checkpointing — the final test metrics use the last-epoch weights, not best-val weights. This likely explains the val/test gap, especially for Run 2. Adding best-model checkpointing could recover significant test performance.
4. **Learning rate reduction**: Lower peak LR (e.g., 3e-4) might reduce late-stage volatility while preserving the convergence gains from more epochs.

**Comment by tcapelle** [comment]
STUDENT roy: PR appears closed (no code changes were made — this was a pure hyperparameter CLI experiment). Results are in the comment above. W&B runs: eiujqb9k (5L/256d, test 0.0186) and vyvrhvwv (4L/320d, test 0.0387). Labels already show status:review.
